keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+used in a security context (if authz check fails, don't show user any other
+info), or more generally to minimize noise where subsequent validation messages
+are of low value if a particular validation fails.
+
+```yaml
+  validations:
+    - expression: "<CEL precondition check>"
+      failureTreatment: IgnoreOtherValidationFailures
+    - expression: "..."
+    - expression: "..."
+```
+
+`failureTreatment` should support:
+
+- `NoMatch` - The resource is considered "not matched" by this policy check and
+  is allowed to be admitted (at least by this policy) regardless of the outcome
+  of other validations.
+- `IgnoreOtherValidationFailures` - Only this policy failure is enforced. Any
+  other validation failures are ignored. If multiple validations have this
+  `failureTreatment` the first in the list to failure is shown and all others
+  are ignored.
+- unset - Apply the configured enforcement for this validation.
+
+Use case: Match GVK instead of GVR
+
+Possible via: 
+- `resourceRules: { apiGroups: "apps", apiVersions: "v1" }`
+- `{expression: "object.metadata.kind in ['Pod', 'Deployment']", failureTreatment: NoMatch}`
+
+(A system that must map from GVK to GVR could also retain a set of well known
+mappings and then fall back to the above only for entries missing from the
+mapping)
+
+#### Metrics
+
+Goals:
+
+- Parity with admission webhook metrics
+  - Should include counter of deny, warn and audit violations
+  - Label by {policy, policy binding, validation expression} identifiers 
+- Counters for number of policy defintions and policy bindings in cluster
+  - Label by state (active vs. error), enforcement action (deny, warn)
+
+Granularity:
+
+Latency metrics should be per {policy, validation expression}. The next level of
+granularity would be {policy, binding, validation expression}, but that
+the number of biindings can become quite large, so let's limit it to
+{policy, validation expression} for now.
+
+- xref: [Metrics Provided by OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/metrics/)
+- xref: [Admission Webhook Metrics](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admission-webhook-metrics)
+
+### User Stories
+
+In addition to "User Stores", see below "Potential Applications" for a list of
+known applications and their use case requirements.
+
+#### Use Case: Singleton Policy
+
+User wishes to define a simple policy that required no parameters. They don't
+want to create parameter CRD since what they're doing can be expressed quite
+simply in a single CEL expression.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  singletonPolicy: true
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
```

Comment:
Consider that the equivalent CEL match expression would probably be 

```go
apiGroups.hasAny(["apps"]) 
&& apiVersions.hasOnly(["v1"]) 
&& operations.hasAny(["CREATE", "UPDATE"]) 
&& resources.hasOnly(["deployments"])
```

You could also treat this as a special function:

```
matchResources({
   "apiGroups": ["apps"],
   "apiVersions": ["v1"]
   "operations":  ["CREATE", "UPDATE"]
   "resources":   ["deployments"]
})
```

Reply:
A recurring theme in this KEP is that there are needs to match or select against data available to K8s which might not be easily expressed as pure logic in CEL. You could create something separate, or you could expose the functionality to CEL. If you expose it to CEL, it's easier to combine with other predicates and will hopefully simplify the number of fields and their possible interactions.

Reply:
Regardless of outcome of https://github.com/kubernetes/enhancements/pull/3492#discussion_r981643590 there might need to be functions in CEL to process matching logic.  For example, there are many uses of the `labelSelector` match object throughout k8s, and having CEL be able to process them could be very useful. I've added a section in phase 2 about CEL function library and add this there.

Reply:
My concern is that you might end up needing a `match` block as it's understood today and a secondary `matchExpresion` to adequately express alternatives among the validation expressions since sometimes what you really want is 'no decision' rather than a trivially true one; however, it can probably wait.

Reply:
This may happen. A lot of webhooks do more fine grained matching in the webhook.

Given that webhooks already have "yaml matching", even if we do improve matching to use more CEL in the future, I don't know if it does much harm to have CEL admission also using "yaml matching". It might be better to keep both as consistent as possible and extend both to support CEL matching if we go that way.

I'm okay with deferring this. I've added a note in the future work section.